### DISCUSSION TITLE
KH inverters no longer support input registers?

### DISCUSSION THREAD

**Discussion by williamjeccles** [nathanmarlor/foxess_modbus#505] [Ideas]
Add support for the FoxESS KH 10.5 series, which currently appears to be unsupported.

![image](https://github.com/nathanmarlor/foxess_modbus/assets/6324545/e17dfb62-efe1-4d38-aa57-3352a0c0c4ab)

**Comment by canton7** [comment]
Ah, they've changed something. We're looking for "KH-" at the start.

**Reply by canton7** [reply]
Opened #506.

**Reply by williamjeccles** [reply]
thanks @canton7 that makes sense, if you ping me the PR I'll test it for you if you like.

**Reply by canton7** [reply]
Cheers, can you give https://github.com/nathanmarlor/foxess_modbus/pull/507 a try?

**Reply by canton7** [reply]
Let me know if you need a beta to test.

**Reply by williamjeccles** [reply]
thanks, working on it now @canton7

**Comment by canton7** [comment]
Fixed in v1.9.0.

**Reply by williamjeccles** [reply]
Hey @canton7, I'm still getting all the values reporting as not available.

I'm pretty sure it's not a cable issue as I've triple-checked the connection between he inverter and the USR-W610.
What is odd is that it responded with it's inverter name in the error before so it must be connecting and talking.
As soon as I used your patched version I can complete install but I can see the following in the logs.

Maybe something about the address start point or slave is different on the KH?

tcp://192.168.1.164:502 247 - 5 failed poll attempts: now not connected. Last error: Error reading registers. Type: RegisterType.INPUT; start: 11000; count: 8; slave: 247 from tcp://192.168.1.164:502: Exception Response(131, 3, IllegalAddress)

**Reply by canton7** [reply]
Interesting. In #171 where we mapped out KH support, we identified both input (11xxx) and holding (31xxx) registers for the KH. I switched to using input registers in v1.5.3b2, which was tested with a KH.

So my best guess is that either the KH you've got has either newer firmware (and FoxESS removed the input register support), or older firmware (before they added input register support). Could you take a look at the manager version please?

**Reply by canton7** [reply]
@MJW911, @lamixer, could you confirm your manager versions as well please?

**Reply by lamixer** [reply]
Hello - I was updated from the firmware my units were shipped with to manager version 1.16 about a month ago. I believe that is the current verison for KH series. Do you want me to poll a register for you? My input registers are still working but I think I set it up before they updated my firmware.

**Reply by lamixer** [reply]
I was on manager 1.11 when we worked on the integration.

**Reply by williamjeccles** [reply]
@lamixer / @canton7 

I'm on the following 
Master : 1.25
Slave : 1.01
Manager : 1.18

**Reply by williamjeccles** [reply]
![image](https://github.com/nathanmarlor/foxess_modbus/assets/6324545/154c2b15-7208-4370-85c0-db0ad72b6744)

**Reply by canton7** [reply]
Thanks both! I wonder whether they've gone and broken the input registers between 1.16 and 1.18 then... That would be annoying, as all of the BMS registers are only available as input registers.

**Reply by williamjeccles** [reply]
@canton7 any easy way to prove this? Happy to manually run a patched version to help test and I can report back to Fox to aid a fix

**Reply by canton7** [reply]
@williamjeccles If you're comfortable using modpoll, that's a quick and easy option. Try reading holding register 31000 and input register 11000.

Alternatively, I've switched the integration to use holding registers for the KH in this PR: https://github.com/nathanmarlor/foxess_modbus/pull/512

**Reply by williamjeccles** [reply]
@canton7 bingo 
![image](https://github.com/nathanmarlor/foxess_modbus/assets/6324545/db10899d-912d-46a2-bd17-1b6b6478bcab)

now working just with some of the BMS and controls for SOC / workmode broken.

Let me know what I should report to Fox unless you have connections in Fox R&D?

**Reply by canton7** [reply]
Cool! 

I don't have any links to Fox R&D, so if you could get some information on whether this is intentional that would be fantastic. If it is, we can switch the integration to use holding registers from a particular manager version onwards.

The information we have is that the KH used to expose input registers as well as holding registers, but somewhere between manager version 1.16 and 1.18 support for input registers seems to have been dropped.

Thanks!

**Reply by williamjeccles** [reply]
@canton7 on it!

**Comment by MJW911** [comment]
sorry folks, was away from emails... ive been using the modbus integration for quite a while now with no issues on my KH10.5. 
master ver: 1.17
slaves: 1.012
manager: 1.13

**Reply by canton7** [reply]
Great, thanks!

**Reply by williamjeccles** [reply]
@MJW911 I can 100% confirm that after 1.18 manager my HA stopped working with input registers and I have to use the patch to switch to holding register types which means I don't have bms or workmode control currently.

I'm currently on 
Master : 1.25
Slave : 1.01
Manager : 1.18

Fox R&D are looking into it also. Likely just an accidental regression that we're the first people to notice.
They've been very responsive so I'm sure it won't be long before a fix is released but will update this thread as things develop.

**Comment by williamjeccles** [comment]


**Reply by canton7** [reply]
Thanks. I don't have facebook messenger on my phone, so I'll check once I'm home this evening.

**Comment by williamjeccles** [comment]
resolved in latest beta version.
